### ISSUE TITLE
Prevent propagation of invalid blocks with block pubsub topic validator

### ISSUE THREAD

**Issue by anorth** [filecoin-project/go-filecoin#2783]
### Description
#2755 reported a network progress issue that ultimately came down to propagation of invalid blocks (which were far ahead of the consensus head #2782).

The spec describes [block validation logic](https://filecoin-project.github.io/specs/docs/validation/) which must execute before a block is accepted or propagated to peers. Validation includes basic well-formedness checks as well as checking that the block can be linked to the nodes existing chain and the contained messages and receipts describe valid state transitions. See @whyrusleeping's [comment on 2755]( https://github.com/filecoin-project/go-filecoin/issues/2755#issuecomment-491590106)

Implement a [libp2p pubsub topic validator](https://github.com/libp2p/go-libp2p-pubsub/blob/master/pubsub.go#L930) which validates blocks before accepting them locally or propagating them to peers. 

It's unclear whether "propagation" in the spec refers to libp2p pubsub validation, or the higher level availability of a block to peers that request it directly. Comprehensive block validation is expensive and requires access to all parent blocks and the preceding state (i.e. fully synced);  doing this in the topic validator may slow block propagation. Alternatively we could split validation into two parts, only one of which applies to the pubsub. We need @whyrusleeping to help us make a call here and, if necessary, update the spec to clarify.

### Acceptance criteria
- Blocks are validated before nodes perform any other processing
- Invalid blocks are not propagated

### Risks + pitfalls
- Tradeoff between block propagation time and validation
- If a block references a parent not already held by the node, it can't validate, or must block validation, until the parent is received.
- The node currently subscribes to the block pubsub before syncing the chain; it won't be able to validate blocks before that so should probably join pubsub only after catching up (#1160)
- The pubsub validator will prevent a node receiving an invalid block on `Subscription.Next()` which will also prevent it storing an explicit list of "bad" blocks, which the spec refers to (but is not yet implemented). Is refusing to accept or propagate a block sufficient? Is caching some validation results in the topic validator sufficient?

**Comment by vyzo** [comment]
Note that we just merged https://github.com/libp2p/go-libp2p-pubsub/pull/176 which reworked the validation pipeline, and there is new pubsub release tag (https://github.com/libp2p/go-libp2p-pubsub/releases/tag/v0.0.3) that includes it.

Another issue you should be aware of is validation time and the interprlay with the timecache that prevents cycles in pubsub.
If your validators are expected to take significant time (more than a couple of seconds), you should adjust the timecache duration by setting `pubsub.TimeCacheDuration`.
The default value is 2 minutes, but you should adjust it to be at least `diameter * validation-time`, where `diameter` is the expected diameter of the network in hops.

**Comment by whyrusleeping** [comment]
@anorth I would say we should for sure ensure the block decodes successfully, the messages line up with the message root, the signature is valid, the timestamp is correct, and the blocks parents are not a known bad tipset. That should be good enough to pass pubsub validation and get kicked off to the chain sync manager.

**Comment by anorth** [comment]
There now exists a `BlockSyntaxValidator` interface, incompletely implemented. This issue now boils down to:
- complete the syntax validator implementation to match the spec
- use the syntax validator in a libp2p topic validator.
